packages/core/contracts/bot-helpers/atomic-liquidator/ReserveCurrencyLiquidator.sol

```diff
@@ -0,0 +1,136 @@
+pragma solidity ^0.6.0;
+pragma experimental ABIEncoderV2;
+
+import "@openzeppelin/contracts/math/SafeMath.sol";
+
+import "@uniswap/lib/contracts/libraries/TransferHelper.sol";
+import "@uniswap/v2-periphery/contracts/interfaces/IUniswapV2Router01.sol";
+
+import "../../common/implementation/FixedPoint.sol";
+
+/**
+ * @title ReserveCurrencyLiquidator
+ * @notice Helper contract to enable a liquidator to hold one reserver currency and liquidate against any number of
+ * financial contracts. Is assumed to be called by a DSProxy which holds reserve currency.
+ */
+
+contract ReserveCurrencyLiquidator {
+    using SafeMath for uint256;
+    using FixedPoint for FixedPoint.Unsigned;
+
+    /**
+     * @notice Swaps required amount of reserve currency to collateral currency which is then used to mint tokens to
+     * liquidate a position within one transaction.
+     * @dev After the liquidation is done the DSProxy that called this method will have an open positon AND pending
+     * liquidation within the financial contract. The bot using the DSProxy should withdraw the liquidation once it has
+     * passed liveness. At this point the position can be manually unwound.
+     * @param uniswapRouter address of the uniswap router used to facilate trades.
+     * @param financialContract address of the financial contract on which the liquidation is occuring.
+     * @param reserveCurrency address of the token to swap for collateral. THis is the common currency held by the DSProxy.
+     * @param liquidatedSponsor address of the sponsor to be liquidated.
+     * @param maxReserverTokenSpent maximum number of reserve tokens to spend in the trade. Bounds slippage.
+     * @param minCollateralPerTokenLiquidated abort the liquidation if the position's collateral per token is below this value.
+     * @param maxCollateralPerTokenLiquidated abort the liquidation if the position's collateral per token exceeds this value.
+     * @param maxTokensToLiquidate max number of tokens to liquidate. For a full liquidation this is the full position debt.
+     * @param deadline abort the trade and liquidation if the transaction is mined after this timestamp.
+     **/
+    function swapMintLiquidate(
+        address uniswapRouter,
+        address financialContract,
+        address reserveCurrency,
+        address liquidatedSponsor,
+        FixedPoint.Unsigned calldata maxReserverTokenSpent,
+        FixedPoint.Unsigned calldata minCollateralPerTokenLiquidated,
+        FixedPoint.Unsigned calldata maxCollateralPerTokenLiquidated,
+        FixedPoint.Unsigned calldata maxTokensToLiquidate,
+        uint256 deadline
+    ) public {
+        IUniswapV2Router01 router = IUniswapV2Router01(uniswapRouter);
+        FinancialContractInterface fc = FinancialContractInterface(financialContract);
+
+        // 1. Calculate how much collateral needed to mint maxTokensToLiquidate.
+        FixedPoint.Unsigned memory gcr = fc.pfc().div(fc.totalTokensOutstanding());
```

Comment:
This makes me think that we should make gcr() a public method on the financial contracts

Reply:
100%. equally, it would be great for the contract to expose methods on computing fees so I dont need both a finder and a store to get this number.

Reply:
Wow, great point. Yeah, we really should expose the GCR in some way.

The return values would ideally help you in computing fees and capital required, but you can't really `.call` those until you _have_ the tokens required or the method will revert 🤔